foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/identitymaps/IdentityMapManager.java

```diff
@@ -488,8 +488,11 @@ public void invalidateQueryCache(Class classThatChanged) {
         }
         Set invalidations = this.queryResultsInvalidationsByClass.get(classThatChanged);
         if (invalidations != null) {
-            for (Object queryKey : invalidations) {
-                this.queryResults.remove(queryKey);
+            synchronized (this.queryResults) {
+                for (Object queryKey : invalidations) {
+                    this.queryResults.remove(queryKey);
```

Comment:
Problem is modification of `invalidations` while it's iterating.  
The code can be rewritten according to your suggestion, but this will not eliminate the need of synchronization. Internally `java.util.concurrent.ConcurrentHashMap.CollectionView#removeAll` can iterate `invalidations` allso.

PS If `invalidations.size() > queryResults.size()` the `CollectionView#removeAll` will work faster then `for (Object queryKey : invalidations) {}`.